File: packages/android-kotlin/src/main/kotlin/com/onchainux/ui/ConnectButton.kt
/**
 * ConnectButton — Jetpack Compose component.
 *
 * A drop-in wallet connection button that adapts its appearance based on
 * connection state. Shows "Connect Wallet" when disconnected, displays
 * the connected address when connected, and shows a loading spinner
 * while connecting.
 *
 * ## Usage
 * ```kotlin
 * ConnectButton()
 * ConnectButton(
 *     label = "Connect Wallet",
 *     variant = ButtonVariant.Secondary,
 *     size = ButtonSize.Large,
 *     showBalance = true,
 *     showAvatar = true
 * )
 * ```
 */
package com.cinacoin.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.TextUnit
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.lifecycle.viewmodel.compose.viewModel
import com.cinacoin.core.AccountInfo
import com.cinacoin.core.ConnectionStatus
import com.cinacoin.core.Cinacoin
import com.cinacoin.core.ThemeColors
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

/** Button size preset. */
enum class ButtonSize(
    val height: Dp,
    val horizontalPadding: Dp,
    val fontSize: TextUnit
) {
    SMALL(height = 36.dp, horizontalPadding = 16.dp, fontSize = 12.sp),
    MEDIUM(height = 44.dp, horizontalPadding = 24.dp, fontSize = 14.sp),
    LARGE(height = 52.dp, horizontalPadding = 32.dp, fontSize = 16.sp)
}

/** Button visual variant. */
enum class ButtonVariant { PRIMARY, SECONDARY, GHOST }

/**
 * Compose Connect Button.
 */
@Composable
fun ConnectButton(
    label: String = "Connect Wallet",
    variant: ButtonVariant = ButtonVariant.PRIMARY,
    size: ButtonSize = ButtonSize.MEDIUM,
    showBalance: Boolean = false,
    showAvatar: Boolean = false,
    showNetwork: Boolean = false,
    onPress: (() -> Unit)? = null,
    onDisconnect: (() -> Unit)? = null,
    onChainUX: Cinacoin = Cinacoin.getInstance(),
    modifier: Modifier = Modifier
) {
    val status by onChainUX.status.collectAsState()
    val account by onChainUX.account.collectAsState()
    val colors = onChainUX.themeColors

    val backgroundColor = when {
        status is ConnectionStatus.CONNECTED || variant == ButtonVariant.SECONDARY ->
            Color(colors.bgCard)
        status is ConnectionStatus.ERROR ->
            Color(colors.error).copy(alpha = 0.15f)
        variant == ButtonVariant.GHOST ->
            Color.Transparent
        else ->
            Color(colors.accent500)
    }

    val foregroundColor = when {
        status is ConnectionStatus.CONNECTED || variant == ButtonVariant.SECONDARY ->
            Color(colors.textPrimary)
        status is ConnectionStatus.ERROR ->
            Color(colors.error)
        variant == ButtonVariant.GHOST ->
            Color(colors.textPrimary)
        else ->
            Color.White
    }

    Box(
        modifier = modifier
            .height(size.height)
            .padding(horizontal = size.horizontalPadding)
            .background(backgroundColor, RoundedCornerShape(24.dp))
            .then(
                if (status is ConnectionStatus.CONNECTED || variant == ButtonVariant.SECONDARY) {
                    Modifier.border(
                        width = 1.dp,
                        color = Color(colors.border),
                        shape = RoundedCornerShape(24.dp)
                    )
                } else {
                    Modifier
                }
            )
            .clickable(enabled = status !is ConnectionStatus.CONNECTING) {
                handleButtonPress(status, account, onChainUX, onPress, onDisconnect)
            },
        contentAlignment = Alignment.Center
    ) {
        when {
            status is ConnectionStatus.CONNECTING -> {
                CircularProgressIndicator(
                    modifier = Modifier.size(20.dp),
                    strokeWidth = 2.dp,
                    color = foregroundColor
                )
            }
            status is ConnectionStatus.CONNECTED && account != null -> {
                Row(
                    horizontalArrangement = Arrangement.spacedBy(8.dp),
                    verticalAlignment = Alignment.CenterVertically
                ) {
                    if (showAvatar) {
                        Box(
                            modifier = Modifier
                                .size(size.fontSize.value.dp)
                                .background(Color(colors.accent500), shape = androidx.compose.foundation.shape.CircleShape)
                        )
                    }
                    Text(
                        text = truncateAddress(account!!.address),
                        fontSize = size.fontSize,
                        color = foregroundColor
                    )
                    if (showBalance) {
                        Text(
                            text = "${account!!.balance} ${account!!.chainSymbol}",
                            fontSize = (size.fontSize.value - 2).sp,
                            color = Color(colors.textSecondary)
                        )
                    }
                    if (showNetwork) {
                        NetworkBadge(chainId = account!!.chainId, colors = colors)
                    }
                }
            }
            status is ConnectionStatus.ERROR -> {
                Text(
                    text = "❌ Error",
                    fontSize = size.fontSize,
                    color = Color(colors.error)
                )
            }
            else -> {
                Text(
                    text = label,
                    fontSize = size.fontSize,
                    color = foregroundColor
                )
            }
        }
    }
}

@Composable
private fun NetworkBadge(chainId: Int, colors: ThemeColors) {
    val name = when (chainId) {
        1 -> "ETH"
        137 -> "POLY"
        42161 -> "ARB"
        56 -> "BSC"
        10 -> "OP"
        else -> chainId.toString()
    }
    Text(
        text = name,
        fontSize = 10.sp,
        color = Color(colors.accent500),
        modifier = Modifier
            .background(Color(colors.accent500).copy(alpha = 0.2f), RoundedCornerShape(4.dp))
            .padding(horizontal = 6.dp, vertical = 2.dp)
    )
}

private fun truncateAddress(address: String, prefix: Int = 4, suffix: Int = 4): String {
    if (address.length <= prefix + suffix + 2) return address
    return "${address.substring(0, prefix + 2)}...${address.substring(address.length - suffix)}"
}

private fun handleButtonPress(
    status: ConnectionStatus,
    account: AccountInfo?,
    onChainUX: Cinacoin,
    onPress: (() -> Unit)?,
    onDisconnect: (() -> Unit)?
) {
    val scope = CoroutineScope(Dispatchers.Main)
    if (status is ConnectionStatus.CONNECTED) {
        scope.launch {
            onChainUX.disconnect()
            onDisconnect?.invoke()
        }
        return
    }
    scope.launch {
        try {
            onChainUX.connect("metamask")
            onPress?.invoke()
        } catch (_: Exception) { /* handled by Cinacoin state */ }
    }
}

File: packages/ios-swift/Sources/OnChainUX/ConnectButton.swift
/**
 * ConnectButton — SwiftUI component.
 *
 * A drop-in wallet connection button that adapts its appearance based on
 * connection state. Shows "Connect Wallet" when disconnected, displays
 * the connected address when connected, and shows a loading spinner
 * while connecting.
 *
 * ## Usage
 * ```swift
 * ConnectButton(label: "Connect Wallet")
 * ConnectButton(size: .large, showBalance: true, showAvatar: true)
 * ```
 */

import SwiftUI

/// Button size preset.
public enum ConnectButtonSize: String, CaseIterable {
    case small = "sm"
    case medium = "md"
    case large = "lg"
    
    var height: CGFloat {
        switch self {
        case .small: return 36
        case .medium: return 44
        case .large: return 52
        }
    }
    
    var horizontalPadding: CGFloat {
        switch self {
        case .small: return 16
        case .medium: return 24
        case .large: return 32
        }
    }
    
    var fontSize: CGFloat {
        switch self {
        case .small: return 12
        case .medium: return 14
        case .large: return 16
        }
    }
}

/// Visual variant of the button.
public enum ConnectButtonVariant: String, CaseIterable {
    case primary
    case secondary
    case ghost
}

/// SwiftUI Connect Button component.
public struct ConnectButton: View {
    @EnvironmentObject private var onChainUX: Cinacoin
    
    /// Button text when disconnected.
    public var label: String
    /// Button visual variant.
    public var variant: ConnectButtonVariant
    /// Button size.
    public var size: ConnectButtonSize
    /// Show account balance when connected.
    public var showBalance: Bool
    /// Show avatar indicator when connected.
    public var showAvatar: Bool
    /// Show network badge when connected.
    public var showNetwork: Bool
    /// Click handler.
    public var onPress: (() -> Void)?
    /// Disconnect handler.
    public var onDisconnect: (() -> Void)?
    
    /// Create a ConnectButton with default settings.
    public init(
        label: String = "Connect Wallet",
        variant: ConnectButtonVariant = .primary,
        size: ConnectButtonSize = .medium,
        showBalance: Bool = false,
        showAvatar: Bool = false,
        showNetwork: Bool = false,
        onPress: (() -> Void)? = nil,
        onDisconnect: (() -> Void)? = nil
    ) {
        self.label = label
        self.variant = variant
        self.size = size
        self.showBalance = showBalance
        self.showAvatar = showAvatar
        self.showNetwork = showNetwork
        self.onPress = onPress
        self.onDisconnect = onDisconnect
    }
    
    public var body: some View {
        Button(action: handlePress) {
            if onChainUX.status == .connecting {
                ProgressView()
                    .progressViewStyle(CircularProgressViewStyle(tint: buttonForegroundColor))
                    .frame(height: size.height)
            } else if case .connected = onChainUX.status, let account = onChainUX.account {
                HStack(spacing: 8) {
                    if showAvatar {
                        Circle()
                            .fill(Color.accentColor)
                            .frame(width: size.fontSize, height: size.fontSize)
                    }
                    Text(truncateAddress(account.address))
                        .font(.system(size: size.fontSize, weight: .medium, design: .monospaced))
                    if showBalance {
                        Text("\(account.balance) \(account.chainSymbol)")
                            .font(.system(size: size.fontSize - 2, weight: .regular))
                            .foregroundColor(onChainUX.themeColors.secondaryColor)
                    }
                    if showNetwork {
                        NetworkBadge(chainId: account.chainId)
                    }
                }
                .foregroundColor(buttonForegroundColor)
            } else if case .error = onChainUX.status {
                Text("❌ Error")
                    .foregroundColor(onChainUX.themeColors.errorColor)
                    .font(.system(size: size.fontSize, weight: .semibold))
            } else {
                Text(label)
                    .foregroundColor(buttonForegroundColor)
                    .font(.system(size: size.fontSize, weight: .semibold))
            }
        }
        .frame(height: size.height)
        .padding(.horizontal, size.horizontalPadding)
        .background(buttonBackgroundColor)
        .cornerRadius(24)
        .overlay(
            RoundedRectangle(cornerRadius: 24)
                .stroke(borderColor, lineWidth: needsBorder ? 1 : 0)
        )
        .disabled(onChainUX.status == .connecting)
        .accessibilityLabel(accessibilityLabel)
    }
    
    // MARK: - Styling
    
    private var buttonBackgroundColor: Color {
        let colors = onChainUX.themeColors
        switch (onChainUX.status, variant) {
        case (.connected, _), (_, .secondary):
            return Color(hex: colors.bgCard)
        case (.error(_), _):
            return Color(hex: colors.error).opacity(0.15)
        case (_, .ghost):
            return Color.clear
        default:
            return Color(hex: colors.accent500)
        }
    }
    
    private var buttonForegroundColor: Color {
        let colors = onChainUX.themeColors
        switch (onChainUX.status, variant) {
        case (.connected, _), (_, .secondary):
            return Color(hex: colors.textPrimary)
        case (.error(_), _):
            return Color(hex: colors.error)
        case (_, .ghost):
            return Color(hex: colors.textPrimary)
        default:
            return .white
        }
    }
    
    private var borderColor: Color {
        let colors = onChainUX.themeColors
        switch (onChainUX.status, variant) {
        case (.connected, _), (_, .secondary):
            return Color(hex: colors.border)
        default:
            return .clear
        }
    }
    
    private var needsBorder: Bool {
        switch (onChainUX.status, variant) {
        case (.connected, _), (_, .secondary):
            return true
        default:
            return false
        }
    }
    
    private var accessibilityLabel: String {
        if case .connected = onChainUX.status, let account = onChainUX.account {
            return "Connected as \(truncateAddress(account.address))"
        }
        return label
    }
    
    // MARK: - Actions
    
    private func handlePress() {
        if case .connected = onChainUX.status {
            Task {
                await onChainUX.disconnect()
                onDisconnect?()
            }
            return
        }
        
        Task {
            do {
                try await onChainUX.connect(connectorId: "metamask")
                onPress?()
            } catch {
                // Error state is handled by Cinacoin
            }
        }
    }
}

/// Network badge showing chain name.
private struct NetworkBadge: View {
    let chainId: Int
    
    var body: some View {
        Text(chainName(for: chainId))
            .font(.system(size: 10, weight: .medium))
            .padding(.horizontal, 6)
            .padding(.vertical, 2)
            .background(Color.accentColor.opacity(0.2))
            .cornerRadius(4)
    }
    
    private func chainName(for chainId: Int) -> String {
        switch chainId {
        case 1: return "ETH"
        case 137: return "POLY"
        case 42161: return "ARB"
        case 56: return "BSC"
        case 10: return "OP"
        default: return String(chainId)
        }
    }
}

// MARK: - Helpers

/// Truncate an Ethereum address for display.
func truncateAddress(_ address: String, prefix: Int = 4, suffix: Int = 4) -> String {
    guard address.count > prefix + suffix + 2 else { return address }
    let startIndex = address.index(address.startIndex, offsetBy: prefix + 2)
    let endIndex = address.index(address.endIndex, offsetBy: -suffix)
    return "\(address.prefix(prefix + 2))...\(address.suffix(suffix))"
}

/// Extension to create Color from hex string.
extension Color {
    init(hex: String) {
        let hex = hex.trimmingCharacters(in: CharacterSet.alphanumerics.inverted)
        var int: UInt64 = 0
        Scanner(string: hex).scanHexInt64(&int)
        let a, r, g, b: UInt64
        switch hex.count {
        case 6: // RGB
            (a, r, g, b) = (255, int >> 16, int >> 8 & 0xFF, int & 0xFF)
        case 8: // ARGB
            (a, r, g, b) = (int >> 24, int >> 16 & 0xFF, int >> 8 & 0xFF, int & 0xFF)
        default:
            (a, r, g, b) = (255, 0, 0, 0)
        }
        self.init(
            .sRGB,
            red: Double(r) / 255,
            green: Double(g) / 255,
            blue: Double(b) / 255,
            opacity: Double(a) / 255
        )
    }
}

extension ThemeColors {
    var secondaryColor: Color { Color(hex: textSecondary) }
    var errorColor: Color { Color(hex: error) }
}
